rpcs3/rpcs3qt/debugger_frame.cpp

```diff
@@ -397,72 +420,127 @@ void debugger_frame::Show_Val()
 	if (cpu)
 	{
 		unsigned long pc = cpu ? GetPc() : 0x0;
-		addr->setText("Address: " + QString("%1").arg(pc, 8, 16, QChar('0')));	// set address input line to 8 digits
-		p_pc->setPlaceholderText(QString("%1").arg(pc, 8, 16, QChar('0')));
+		address_preview_label->setText("Address: " + QString("0x%1").arg(pc, 8, 16, QChar('0')));
+		expression_input->setPlaceholderText(QString("0x%1").arg(pc, 8, 16, QChar('0')));
 	}
 	else
 	{
-		p_pc->setPlaceholderText("00000000");
-		addr->setText("Address: 00000000");
+		expression_input->setPlaceholderText("0x00000000");
+		address_preview_label->setText("Address: 0x00000000");
 	}
 
 	auto l_changeLabel = [=]()
 	{
-		if (p_pc->text().isEmpty())
+		if (expression_input->text().isEmpty())
 		{
-			addr->setText("Address: " + p_pc->placeholderText());
+			address_preview_label->setText("Address: " + expression_input->placeholderText());
 		}
 		else
 		{
-			bool ok;
-			ulong ul_addr = p_pc->text().toULong(&ok, 16);
-			addr->setText("Address: " + QString("%1").arg(ul_addr, 8, 16, QChar('0'))); // set address input line to 8 digits
+			ulong ul_addr = EvaluateExpression(expression_input->text());
+			address_preview_label->setText("Address: " + QString("0x%1").arg(ul_addr, 8, 16, QChar('0')));
 		}
 	};
 
-	connect(p_pc, &QLineEdit::textChanged, l_changeLabel);
+	connect(expression_input, &QLineEdit::textChanged, l_changeLabel);
 	connect(button_ok, &QAbstractButton::clicked, diag, &QDialog::accept);
 	connect(button_cancel, &QAbstractButton::clicked, diag, &QDialog::reject);
 
 	diag->move(QCursor::pos());
 
 	if (diag->exec() == QDialog::Accepted)
 	{
-		unsigned long pc = cpu ? GetPc() : 0x0;
-		if (p_pc->text().isEmpty())
+		u32 address = cpu ? GetPc() : 0x0;
+
+		if (expression_input->text().isEmpty())
 		{
-			addr->setText(p_pc->placeholderText());
+			address_preview_label->setText(expression_input->placeholderText());
 		}
 		else
 		{
-			bool ok;
-			pc = p_pc->text().toULong(&ok, 16);
-			addr->setText(p_pc->text());
+			address = EvaluateExpression(expression_input->text());
+			address_preview_label->setText(expression_input->text());
 		}
-		m_list->ShowAddr(CentrePc(pc));
+
+		m_list->ShowAddress(address);
 	}
 
 	diag->deleteLater();
 }
 
+u64 debugger_frame::EvaluateExpression(const QString& expression)
+{
+	auto thread = cpu.lock();
+
+	// Parse expression
+	QJSEngine scriptEngine;
+	scriptEngine.globalObject().setProperty("pc", GetPc());
+
+	if (thread->id_type() == 1)
+	{
+		auto ppu = static_cast<ppu_thread*>(thread.get());
+
+		for (int i = 0; i < 32; ++i)
+			scriptEngine.globalObject().setProperty(QString("r%1").arg(i), QJSValue((double)(ppu->gpr[i])));
+
+		scriptEngine.globalObject().setProperty("lr", QJSValue((double)ppu->lr));
```

Comment:
Double type cannot encode full range of 64-bit unsigned value. Is it really what you meant?

Reply:
There is no alternative